### PR TITLE
feat(api): subdomain support

### PR DESCRIPTION
**Pull request by aloknerurkar**
### Checklist

- [ ] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md)
- [ ] My change requires a documentation update and I have done it
- [ ] I have added tests to cover my changes.

### Screenshots (if appropriate):

---
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/2970)

### CODE REVIEW

**Review by acud — CHANGES_REQUESTED** [review]
very nicely done!

> Reviewed 5 of 9 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2970#-:-N2qX40wEMI5mWXs7AyO:b-1znowi)* status: all files reviewed, 3 unresolved discussions (waiting on @aloknerurkar)

___
*[`pkg/api/router.go` line 42 at r4](https://reviewable.io/reviews/ethersphere/bee/2970#-N2qUa2J0k6veDPwBFMh:-N2qUa2J0k6veDPwBFMi:boy7wj0) ([raw file](https://github.com/ethersphere/bee/blob/85dcc99a578001d6174a931b0680aea43c42d8f8/pkg/api/router.go#L42)):*
> ```Go
> 
> 	subdomainRouter := router.Host("{subdomain:.*}.localhost").Subrouter()
> ```

i think that with @agazso we discussed having the locators in the form of `*.bzz OR swarm.localhost/` i.e. to prefix everything with the subdomain of localhost, like ipfs are doing. not sure how to handle this here.
___
*[`pkg/api/subdomain.go` line 1 at r4](https://reviewable.io/reviews/ethersphere/bee/2970#-N2qViu57a5cLq27bA8p:-N2qViu57a5cLq27bA8q:bldwpal) ([raw file](https://github.com/ethersphere/bee/blob/85dcc99a578001d6174a931b0680aea43c42d8f8/pkg/api/subdomain.go#L1)):*
> ```Go
> package api
> ```

code copy missing on new files
___
*[`pkg/resolver/multiresolver/multiresolver.go` line 106 at r4](https://reviewable.io/reviews/ethersphere/bee/2970#-N2qWxBf4xg5xnpUe3XF:-N2qWxBf4xg5xnpUe3XG:b1zedqa) ([raw file](https://github.com/ethersphere/bee/blob/85dcc99a578001d6174a931b0680aea43c42d8f8/pkg/resolver/multiresolver/multiresolver.go#L106)):*
> ```Go
> }
> 
> func WithDefaultStaticResolver(staticResolver resolver.Interface) Option {
> ```

with cid resolver?

**Comment by aloknerurkar** [comment]
*[`pkg/api/router.go` line 42 at r4](https://reviewable.io/reviews/ethersphere/bee/2970#-N2qUa2J0k6veDPwBFMh:-N2rjILgBI29JB_FtKLw:bgap3wo) ([raw file](https://github.com/ethersphere/bee/blob/85dcc99a578001d6174a931b0680aea43c42d8f8/pkg/api/router.go#L42)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

i think that with @agazso we discussed having the locators in the form of `*.bzz OR swarm.localhost/` i.e. to prefix everything with the subdomain of localhost, like ipfs are doing. not sure how to handle this here.
</blockquote></details>

So it seems to be as simple as adding the string.

**Review by aloknerurkar — COMMENTED** [review]
It seems to be as simple as using the string in the regex. We can easily change it later.

> Dismissed @acud from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2970#-:-N2rjago0FS1lZ8mDvC6:b-wxytqe)* status: 6 of 11 files reviewed, 1 unresolved discussion (waiting on @acud, @agazso, @janos, and @mrekucci)

___
*[`pkg/api/subdomain.go` line 1 at r4](https://reviewable.io/reviews/ethersphere/bee/2970#-N2qViu57a5cLq27bA8p:-N2rjT773JU_-wiOFey6:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/85dcc99a578001d6174a931b0680aea43c42d8f8/pkg/api/subdomain.go#L1)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

code copy missing on new files
</blockquote></details>

Done.
___
*[`pkg/resolver/multiresolver/multiresolver.go` line 106 at r4](https://reviewable.io/reviews/ethersphere/bee/2970#-N2qWxBf4xg5xnpUe3XF:-N2rjYkI1E9zELxO93Os:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/85dcc99a578001d6174a931b0680aea43c42d8f8/pkg/resolver/multiresolver/multiresolver.go#L106)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

with cid resolver?
</blockquote></details>

Done.

File: pkg/api/bzz.go [hunk 1]
```diff
@@ -260,6 +256,16 @@ func (s *server) bzzDownloadHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	s.serveReference(address, pathVar, w, r)
+}
+
+func (s *server) serveReference(address swarm.Address, pathVar string, w http.ResponseWriter, r *http.Request) {
```
Comment (aloknerurkar) [inline pkg/api/bzz.go:262]: @mrekucci So this method is now a helper which can be called from different callers. Ideally we would want to have the logs tagged with the parent request, so basically distinguish if we are serving on `/bzz` or on subdomain. Would this be something handled better with you new logging handler?

Right now we can pass a logger instance to this method which can be prefixed with some string based on the caller. But then we would need to have some basic proxy implementation of the logger to add the prefix.

File: pkg/api/subdomain.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	pathVar := mux.Vars(r)["path"]
+	if strings.HasSuffix(pathVar, "/") {
+		pathVar = strings.TrimRight(pathVar, "/")
+		// NOTE: leave one slash if there was some
+		pathVar += "/"
+	}
+
+	address, err := s.resolveNameOrAddress(nameOrHex)
+	if err != nil {
+		logger.Debugf("bzz download: parse address %s: %v", nameOrHex, err)
```
Comment (aloknerurkar) [inline pkg/api/subdomain.go:29]: change these log messages

**Review by acud — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2970#-:-N2uGgPMBuo0_CyMVZnZ:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @aloknerurkar, @janos, and @mrekucci)

___
*[`pkg/api/subdomain.go` line 29 at r5](https://reviewable.io/reviews/ethersphere/bee/2970#-N2STwWT0Ro11Z6p4yoe-r5-29:-N2uGAmICEt1M9QEyEBF:b7r74o) ([raw file](https://github.com/ethersphere/bee/blob/975fad8058c842fce2efba455dd77b3c7efbf158/pkg/api/subdomain.go#L29)):*
<details><summary><i>Previously, aloknerurkar wrote…</i></summary><blockquote>

change these log messages
</blockquote></details>

note to self? :)
___
*[`pkg/resolver/multiresolver/multiresolver.go` line 107 at r5](https://reviewable.io/reviews/ethersphere/bee/2970#-N2uG_IT6_d1ez6IEayY:-N2uG_IT6_d1ez6IEayZ:b-mh924b) ([raw file](https://github.com/ethersphere/bee/blob/975fad8058c842fce2efba455dd77b3c7efbf158/pkg/resolver/multiresolver/multiresolver.go#L107)):*
> ```Go
> }
> 
> func WithDefaultCIDResolver() Option {
> ```

nit, would also get rid of the `Default` notion here

**Review by aloknerurkar — COMMENTED** [review]
> Dismissed @acud from a discussion.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2970)* status: all files reviewed, 4 unresolved discussions (waiting on @acud, @janos, and @mrekucci)

___
*[`pkg/api/subdomain.go` line 29 at r5](https://reviewable.io/reviews/ethersphere/bee/2970#-N2STwWT0Ro11Z6p4yoe-r5-29:-N2uIP-X8SQpMIqHWeh-:bwq0zg9) ([raw file](https://github.com/ethersphere/bee/blob/975fad8058c842fce2efba455dd77b3c7efbf158/pkg/api/subdomain.go#L29)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

note to self? :)
</blockquote></details>

It is one of those things that dont warrant a new commit :p!
___
*[`pkg/resolver/multiresolver/multiresolver.go` line 107 at r5](https://reviewable.io/reviews/ethersphere/bee/2970#-N2uG_IT6_d1ez6IEayY:-N2uIb0ZBRZKfH4Z_KDx:b-k73nfo) ([raw file](https://github.com/ethersphere/bee/blob/975fad8058c842fce2efba455dd77b3c7efbf158/pkg/resolver/multiresolver/multiresolver.go#L107)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

nit, would also get rid of the `Default` notion here
</blockquote></details>

So the reason I kept the 'Default' is that there is this concept of default chain in this multiresolver. So basically if there is no `tld` specified it is configured as default on map key `""`. I think the name is not that big so should be okay.

File: pkg/resolver/cidv1/cidv1.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
+package cidv1
```
Comment (janos) [inline pkg/resolver/cidv1/cidv1.go:1]: Missing copyright header.

File: pkg/resolver/cidv1/cidv1_test.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
+package cidv1_test
```
Comment (janos) [inline pkg/resolver/cidv1/cidv1_test.go:1]: Missing copyright header.

**Review by janos — APPROVED** [review]
Very nice!

**Review by mrekucci — APPROVED** [review]
> Reviewed 3 of 9 files at r1, 1 of 3 files at r2, 4 of 5 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2970)* status: all files reviewed, 6 unresolved discussions (waiting on @acud, @aloknerurkar, @janos, and @mrekucci)

___
*[`pkg/api/api_test.go` line 126 at r6](https://reviewable.io/reviews/ethersphere/bee/2970#-N2vAang2d3sEItCy36p:-N2vAanh33whnAwjibXA:b-l0otqo) ([raw file](https://github.com/ethersphere/bee/blob/17b154b0fcd0829cf1cc2dc308bfa8704c3389d9/pkg/api/api_test.go#L126)):*
> ```Go
> 			Transport: web.RoundTripperFunc(func(r *http.Request) (*http.Response, error) {
> 				requestURL := r.URL.String()
> 				if !strings.HasPrefix(requestURL, "http://") {
> ```

You can do it more efficiently by leveraging the `Scheme` field: `r.URL.Scheme == "http"`.
___
*[`pkg/api/router.go` line 43 at r6](https://reviewable.io/reviews/ethersphere/bee/2970#-N2vDJa27EzJ_qeOfUjO:-N2vDJa27EzJ_qeOfUjP:b-jk5ke8) ([raw file](https://github.com/ethersphere/bee/blob/17b154b0fcd0829cf1cc2dc308bfa8704c3389d9/pkg/api/router.go#L43)):*
> ```Go
> 	subdomainRouter := router.Host("{subdomain:.*}.swarm.localhost").Subrouter()
> 
> 	subdomainRouter.Handle("/{path:.*}", jsonhttp.MethodHandler{
> ```

This endpoint should also be reflected in the OpenAPI spec.

**Review by aloknerurkar — COMMENTED** [review]
> Dismissed @janos from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/2970)* status: 9 of 11 files reviewed, 2 unresolved discussions (waiting on @acud, @janos, and @mrekucci)

___
*[`pkg/api/router.go` line 43 at r6](https://reviewable.io/reviews/ethersphere/bee/2970#-N2vDJa27EzJ_qeOfUjO:-N3-8vOD85NhULHX_-jr:b7yvwbw) ([raw file](https://github.com/ethersphere/bee/blob/17b154b0fcd0829cf1cc2dc308bfa8704c3389d9/pkg/api/router.go#L43)):*
<details><summary><i>Previously, mrekucci wrote…</i></summary><blockquote>

This endpoint should also be reflected in the OpenAPI spec.
</blockquote></details>

So I was going through the OpenAPI documentation to see how to add this. From my understanding we may have to create a separate yaml file for this. The only way I found is to add a server field. But then there is no way to say which paths are applicable for which server. The servers are mainly used to run the same APIs on different environments. 

In our case we need a way to specify an API host and then somehow change the list of allowed paths based on different API hosts. If you are aware of any examples of how this could be done let me know. If not we will merge this with some release notes stating this as an experimental feature which can be opened for testing. Hopefully we can figure out how we can document this by then.
___
*[`pkg/resolver/cidv1/cidv1.go` line 1 at r5](https://reviewable.io/reviews/ethersphere/bee/2970#-N2STwWT0Ro11Z6p4yoh-r5-1:-N3-AfnpCf15Uk3LlWSI:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/975fad8058c842fce2efba455dd77b3c7efbf158/pkg/resolver/cidv1/cidv1.go#L1)):*
<details><summary><i>Previously, janos (Janoš Guljaš) wrote…</i></summary><blockquote>

Missing copyright header.
</blockquote></details>

Done.
___
*[`pkg/resolver/cidv1/cidv1_test.go` line 1 at r5](https://reviewable.io/reviews/ethersphere/bee/2970#-N2STwWT0Ro11Z6p4yoi-r5-1:-N3-AgkN4MotRrFHcTR_:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/975fad8058c842fce2efba455dd77b3c7efbf158/pkg/resolver/cidv1/cidv1_test.go#L1)):*
<details><summary><i>Previously, janos (Janoš Guljaš) wrote…</i></summary><blockquote>

Missing copyright header.
</blockquote></details>

Done.

**Pull request merged**